### PR TITLE
Planet express paging

### PR DESCRIPTION
**Pull request by VsevolodShpagin**

### CODE REVIEW

File: team_planet_express/src/main/java/console_ui/item/paging/PagingUIElement.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
+package console_ui.item.paging;
+
+import console_ui.UserCommunication;
+import core.support.paging.PageNavigation;
+import core.support.paging.PagingRule;
+
+public class PagingUIElement {
```
[the remark this answers is not in the document]
Comment (arthurkov98) [reply team_planet_express/src/main/java/console_ui/item/paging/PagingUIElement.java:7]: Idk. I understand why you do this but or it must be very simple(Victor's example) or must be made with according our architecture, principle and other useful advise.

**Review by arthurkov98 — APPROVED** [review]

File: team_planet_express/src/main/java/console_ui/actions/shared/SearchItemUIAction.java [hunk 2]
```diff
@@ -21,19 +20,19 @@ public class SearchItemUIAction extends UIAction {
...
     private static final String MESSAGE_NO_MATCH = "No items matched search parameters.";
     private static final String MESSAGE_SEARCH_RESULTS = "Search results:";
 
     private final SearchItemService searchItemService;
+    private final OrderingUIElement orderingUIElement;
+    private final PagingUIElement pagingUIElement;
     private final UserCommunication userCommunication;
 
-    public SearchItemUIAction(SearchItemService searchItemService, UserCommunication userCommunication) {
+    public SearchItemUIAction(SearchItemService searchItemService, OrderingUIElement orderingUIElement, PagingUIElement pagingUIElement, UserCommunication userCommunication) {
```
[the remark this answers is not in the document]
Comment (VsevolodShpagin) [reply team_planet_express/src/main/java/console_ui/actions/shared/SearchItemUIAction.java:31]: as far as I know, those can't be mocked, if they are local. therefore they are not local.

File: team_planet_express/src/main/java/console_ui/item/paging/PagingUIElement.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                pageNumberDelta = -1;
+            } else if (PageNavigation.EXIT.getText().equalsIgnoreCase(userInput)) {
+                return false;
+            }
+        } while (!isUserInputValid(pageNumberDelta));
+        changePage(pagingRule, pageNumberDelta);
+        return true;
+    }
+
+    private String getPageNavigationOptions(Integer pageNumber, Integer pageSize, Integer totalFoundItemCount) {
```
[the remark this answers is not in the document]
Comment (VsevolodShpagin) [reply team_planet_express/src/main/java/console_ui/item/paging/PagingUIElement.java:52]: what exactly is insane and in what way?

File: team_planet_express/src/main/java/console_ui/item/paging/PagingUIElement.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (VsevolodShpagin) [reply team_planet_express/src/main/java/console_ui/item/paging/PagingUIElement.java:7]: but it does nothing related to business logic. what data does it process? search results are still created by the service.
this is just concerned with cycling through pages.
you want to add request/response there or something?
we have UIMenu that checks input action numbers. this also just checks ui input, that has nothing to do with business logic, isnt' it?

File: team_planet_express/src/main/java/console_ui/actions/shared/SearchItemUIAction.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (VsevolodShpagin) [reply team_planet_express/src/main/java/console_ui/actions/shared/SearchItemUIAction.java:31]: also, we might want to use this in ListShopItems, for example.

File: team_planet_express/src/main/java/core/responses/shared/SearchItemResponse.java [hunk 4]
```diff
@@ -8,15 +8,24 @@
 
 public class SearchItemResponse extends CoreResponse {
 
-    private final List<Item> items;
+    private List<Item> items;
+    private Integer totalFoundItemCount;
```
[the remark this answers is not in the document]
Comment (VsevolodShpagin) [reply team_planet_express/src/main/java/core/responses/shared/SearchItemResponse.java:12]: how is it useless, if it is used in the paging part?

File: team_planet_express/src/main/java/core/services/item/ordering/OrderingService.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import core.domain.item.Item;
+import core.support.ordering.OrderBy;
+import core.support.ordering.OrderDirection;
+import core.support.ordering.OrderingRule;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+
+public class OrderingService {
```
[the remark this answers is not in the document]
Comment (VsevolodShpagin) [reply team_planet_express/src/main/java/core/services/item/ordering/OrderingService.java:12]: we were told, that static things are bad.
also, it is created only once anyway, what's the point of making it static?
also, also, static things are unmockable.

File: team_planet_express/src/main/java/console_ui/actions/shared/SearchItemUIAction.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (arthurkov98) [reply team_planet_express/src/main/java/console_ui/actions/shared/SearchItemUIAction.java:31]: Action contains element. It is its work init and setup elements.

[the remark this answers is not in the document]
Comment (VsevolodShpagin) [reply team_planet_express/src/main/java/console_ui/actions/shared/SearchItemUIAction.java:31]: the same way as all our services contain validators..

File: team_planet_express/src/main/java/console_ui/item/paging/PagingUIElement.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (VsevolodShpagin) [reply team_planet_express/src/main/java/console_ui/item/paging/PagingUIElement.java:7]: how can we ensure that user input is valid without looping?
it is the same situation, as with action number input.

File: team_planet_express/src/main/java/core/services/actions/shared/SearchItemService.java [hunk 6]
```diff
@@ -5,36 +5,39 @@
...
+    private final OrderingService orderingService;
+    private final PagingService pagingService;
 
-    public SearchItemService(Database database, SearchItemValidator validator) {
+    public SearchItemService(Database database, SearchItemValidator validator, OrderingService orderingService, PagingService pagingService) {
         this.database = database;
         this.validator = validator;
+        this.orderingService = orderingService;
+        this.pagingService = pagingService;
     }
```
[the remark this answers is not in the document]
Comment (arthurkov98) [reply team_planet_express/src/main/java/core/services/actions/shared/SearchItemService.java:28]: Computing module must be static.

[the remark this answers is not in the document]
Comment (VsevolodShpagin) [reply team_planet_express/src/main/java/core/services/actions/shared/SearchItemService.java:28]: what benefits does it provide?
it's not, like we are creating 100 of that class's instances..
it just makes those unmockable.

[the remark this answers is not in the document]
Comment (VsevolodShpagin) [reply team_planet_express/src/main/java/core/services/actions/shared/SearchItemService.java:28]: also, Java does not allow this to be static to begin with..

**Pull request merged**